### PR TITLE
NSX-T Network backing of VM creation

### PR DESCRIPTION
**Pull request by hashneo**
## Description of change

When you bootstrapped a vSphere instance specified the external and primary networks on an NSX-T network port, Juju ignored the Opaque Network type and assumed it was a plain switch creating a VM using this backing type. vSphere would create dummy network port and the bootstrapping would be disconnected from the network.

## QA steps

*Please replace with how we can verify that the change works?*

```sh
juju bootstrap home vsp-controller --default-model lab --config primary-network=<nsx-t int net> --config external-network=<nsx-t ext net> --config datastore=<ds> 
```

## Documentation changes

none

## Bug reference

none

### CODE REVIEW

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 1]
```diff
@@ -518,12 +518,30 @@ func (c *Client) addNetworkDevice(
...
+
+		switch networkSummary.Network.Type {
+		case "OpaqueNetwork":
+			// create an opaque network type
+			var opaqueSummary = network.Summary
+
+			networkBacking = &types.VirtualEthernetCardOpaqueNetworkBackingInfo{
+				OpaqueNetworkId:   opaqueSummary.(*types.OpaqueNetworkSummary).OpaqueNetworkId,
+				OpaqueNetworkType: opaqueSummary.(*types.OpaqueNetworkSummary).OpaqueNetworkType,
+			}
```
Comment (SimonRichardson) [inline provider/vsphere/internal/vsphereclient/createvm.go:533]:
```suggestion
			opaqueSummary, ok := network.Summary.(*types.OpaqueNetworkSummary)
			if !ok {
				return nil, errors.NotValidf("network summary %q", networkSummary.Network.Type)
			}
			networkBacking = &types.VirtualEthernetCardOpaqueNetworkBackingInfo{
				OpaqueNetworkId:   opaqueSummary.OpaqueNetworkId,
				OpaqueNetworkType: opaqueSummary.OpaqueNetworkType,
			}
```

File: provider/vsphere/internal/vsphereclient/createvm.go [hunk 2]
```diff
@@ -518,12 +518,30 @@ func (c *Client) addNetworkDevice(
...
+		case "Network":
+			// create a backing info for a plain old network interface.
+			networkBacking = &types.VirtualEthernetCardNetworkBackingInfo{
+				VirtualDeviceDeviceBackingInfo: types.VirtualDeviceDeviceBackingInfo{
+					DeviceName: network.Name,
+				},
+			}
+			break
+		default:
+			return nil, errors.NewNotValid(nil, fmt.Sprintf("Unknown network type %q", networkSummary.Network.Type))
```
Comment (SimonRichardson) [inline provider/vsphere/internal/vsphereclient/createvm.go:544]:
```suggestion
			return nil, errors.NotValidf("Unknown network type %q", networkSummary.Network.Type)
```

**Review by SimonRichardson — CHANGES_REQUESTED** [review]
Can we get some tests around this, so we don't break it in the future.

**Comment by petevg** [comment]
Closing to clean up old PRs. Feel free to re-open when/if someone returns to this!

**Pull request closed without merging**